lib/web_ui/lib/src/engine/pointer_converter.dart

```diff
@@ -0,0 +1,255 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+part of engine;
+
+class _PointerState {
+  _PointerState(this.x, this.y);
+
+  int get pointer => _pointer; // The identifier used in framework hit test.
+  int _pointer;
+  static int _pointerCount = 0;
+  void startNewPointer() {
+    _pointerCount += 1;
+    _pointer = _pointerCount;
+  }
+
+  bool down = false;
+
+  double x;
+  double y;
+}
+
+/// Converter to convert web pointer data into a form that framework can
+/// understand.
+///
+/// This converter calculates pointer location delta and pointer identifier for
+/// each pointer. Both are required by framework to correctly trigger gesture
+/// activity. It also attempts to sanitize pointer data input sequence by always
+/// synthesizing an add pointer data prior to hover or down if it the pointer is
+/// not previously added.
+///
+/// For example:
+///   before:
+///     hover -> down -> move -> up
+///   after:
+///     add(synthesize) -> hover -> down -> move -> up
+///
+///   before:
+///     down -> move -> up
+///   after:
+///     add(synthesize) -> down -> move -> up
+class PointerDataConverter {
+  PointerDataConverter();
+
+  // Map from browser pointer identifiers to PointerEvent pointer identifiers.
+  final Map<int, _PointerState> _pointers = <int, _PointerState>{};
+
+  /// Clears the existing pointer states.
+  ///
+  /// This method is invoked during hot reload to make sure we have a clean
+  /// converter after hot reload.
+  void clearPointerState() {
+    _pointers.clear();
+    _PointerState._pointerCount = 0;
+  }
+
+  _PointerState _ensureStateForPointer(ui.PointerData datum, double x, double y) {
+    return _pointers.putIfAbsent(
+      datum.device,
+        () => _PointerState(x, y),
+    );
+  }
+
+  ui.PointerData _generateCompletePointerData(ui.PointerData datum) {
+    assert(_pointers.containsKey(datum.device));
+    final _PointerState state = _pointers[datum.device];
+    final double deltaX = datum.physicalX - state.x;
+    final double deltaY = datum.physicalY - state.y;
+    state.x = datum.physicalX;
+    state.y = datum.physicalY;
+    return ui.PointerData(
```

Comment:
If I understand this correctly, we are basically doubling the number of `ui.PointerData` instances that we create. Would it be possible to do this without the extra allocations?

Reply:
ui.PointerData is not mutable, and I don't think we should change it consider this has to be mirror to lib/ui/pointer.dart as well.

If we want to save the allocations, we will have to rewrite pointer_binding. Now I think about it, rewriting might be a better approach..

Reply:
I'm not sure if this is a good idea or not, but here it is:

We could have a function `createPointerData()` that takes all the arguments that `ui.PointerData` takes, and calculates the delta, etc. Then you can call this function in `pointer_binding` to create all pointer data instances.

Reply:
fixed